### PR TITLE
Extend clone macro

### LINKED ISSUE #532: Default behavior of the clone macro
I just realized something: by default, we still clone the content, not using the downgrade/upgrade mechanisms. From my point of view, we should do the opposite: downgrading/upgrading by default or cloning by specifying `@clone`.

cc @sdroege @EPashkin @antoyo

### LINKED ISSUE #533: panic as clone return value?
In some cases, if the upgrading fails, we'd like to panic. For now it looks like this:

```rust
let listbox = gtk::ListBox::new();
listbox.bind_model(Some(&model),
    clone!(@weak window => @default-return panic!("failed to upgrade"), move |item| {
        // blablabla
    }));
```

What about adding another macro to handle such case? Like `clone_or_panic!` or something along the line?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #532.
Fixes #533.

cc @sdroege @EPashkin @antoyo

### CODE REVIEW

**Comment by sdroege** [comment]
You might also want to add another branch to the macro after the `$variable:ident` case for `$variable:expr`, which then maps to `compile_error!` for giving a better error message for things like `x.y` as variable.

Did you test that all the other parts are working? The syntax looks fine to me in any case. Maybe we can add some tests for this (and fix the existing tests)? :)

**Comment by GuillaumeGomez** [comment]
`cargo test` is running fine on my computer?

**Comment by GuillaumeGomez** [comment]
Oh you meant the examples maybe? Well, the PR is already open, just not up to date.

**Comment by GuillaumeGomez** [comment]
I also added a check for the keyword (better than having a macro error. :) )

File: src/clone.rs [hunk 1]
```diff
@@ -223,12 +227,24 @@ macro_rules! to_return_value {
...
 ///
-///         let closure = clone!(self => move |x| {
+///         let closure = clone!(@strong self => move |x| {
 ///             println!("v: {}, x: {}", v, x);
 ///         });
 ///     }
 /// }
 /// ```
+///
+/// **Using another keyword than `weak` or `strong`**:
```
Comment (sdroege) [inline src/clone.rs]: Why do you put an example in the docs for this? :) Nobody should ever do that and having that in the docs only distracts really.

**Comment by sdroege** [comment]
@GuillaumeGomez Improved the error handling a bit and also fixed a bug. This should be good to go I guess :)

File: src/clone.rs [hunk 2]
```diff
@@ -69,33 +69,40 @@ impl<T> Upgrade for rc::Weak<T> {
...
+        compile_error!("You need to specify if this is a weak or a strong clone.");
+    );
+    (@strong $variable:ident) => (
         let $variable = $variable.clone();
     );
     (@weak $variable:ident) => (
         let $variable = $crate::clone::Downgrade::downgrade(&$variable);
     );
+    (@ $keyword:ident $variable:ident) => (
+        compile_error!("Unknown keyword, only `weak` and `strong` are allowed");
```
Comment (GuillaumeGomez) [inline src/clone.rs]: The `stringify!(keyword)` wasn't working?

**Comment by sdroege** [comment]
>The `stringify!(keyword)` wasn't working?

It was but compile_error! does not allow formatting currently, it requires a single string.

Maybe you can get around that with concat! etc, give it a try :)

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by GuillaumeGomez** [comment]
Na, not worth it. Looks good to me now. Once merged, I'll update my examples PR.

**Comment by sdroege** [comment]
Merged, please fix the examples then :) Thanks!

**Comment by GuillaumeGomez** [comment]
Will do before the end of the day!

**Pull request merged**